src/core/image.cpp

```diff
@@ -4421,6 +4421,15 @@ void Image::Deallocate( ) {
         fftwf_destroy_plan(plan_bwd);
         planned = false;
     }
+
+#ifdef ENABLEGPU
+    UnRegisterPageLockedMemory( );
+    if ( device_image_ptr ) {
```

Comment:
This is always null, right? I could not find a place where this gets set to a value....

Reply:
That's a good catch, and since things work with my oversight, I think it reveals something important:

The entire reason that a GpuImage had to know about a specific Image instance (the reason to add these pointers) was that the GpuImage was managing the properties of that Imgae instance. I.e. whether and when the Image data resource was page locked or not.

By returning that data management responsibility to the data-owning object, the need for the GpuImage to know anything about the Image directly is obviated. Now, initialization of a GpuImage based on an Image may be reduced to a simple convenience. On a transfer, the meta-data can be compared in a DebugAssert, and the Image::RegisterPageLockedMemory() can be called with impunity as it will be a no-op if already locked.

I'll work up the changes and also add some test cases to SamplesFunctionalTesting to make sure I didn't miss anything else.

Thankyou!